### PR TITLE
DEPR: Enforce deprecation of na_sentinel

### PR DESCRIPTION
**Pull request by rhshadrach**
- [ ] closes #xxxx (Replace xxxx with the GitHub issue number)
- [x] [Tests added and passed](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#writing-tests) if fixing a bug or adding a new feature
- [x] All [code checks passed](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#pre-commit).
- [x] Added [type annotations](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#type-hints) to new arguments/methods/functions.
- [x] Added an entry in the latest `doc/source/whatsnew/vX.X.X.rst` file if fixing a bug or adding a new feature.

cc @jbrockmendel @jorisvandenbossche

### CODE REVIEW

File: doc/source/whatsnew/v2.0.0.rst [hunk 1]
```diff
@@ -277,7 +277,7 @@ Removal of prior version deprecations/changes
 - Removed the deprecated ``base`` and ``loffset`` arguments from :meth:`pandas.DataFrame.resample`, :meth:`pandas.Series.resample` and :class:`pandas.Grouper`. Use ``offset`` or ``origin`` instead (:issue:`31809`)
 - Changed behavior of :meth:`DataFrame.any` and :meth:`DataFrame.all` with ``bool_only=True``; object-dtype columns with all-bool values will no longer be included, manually cast to ``bool`` dtype first (:issue:`46188`)
 - Enforced deprecation of silently dropping columns that raised a ``TypeError`` in :class:`Series.transform` and :class:`DataFrame.transform` when used with a list or dictionary (:issue:`43740`)
--
+- Removed ``na_sentinel`` argument from :func:`factorize`, :meth:`.Index.factorize`, and :meth:`.ExtensionArray.factorize` (:issue:`46910`)
```
Comment (mroeschke) [inline doc/source/whatsnew/v2.0.0.rst:280]:
```suggestion
- Removed ``na_sentinel`` argument from :func:`factorize`, :meth:`.Index.factorize`, and :meth:`.ExtensionArray.factorize` (:issue:`47157`)
```

I think we're usually referring to the PR that introduced the deprecation

**Review by mroeschke — APPROVED** [review]
Nice cleanup. LGTM

**Comment by mroeschke** [comment]
@rhshadrach heads up - there's a merge conflict

**Comment by rhshadrach** [comment]
Thanks @mroeschke, it's now resolved.

**Comment by jbrockmendel** [comment]
circleCI failure looks like its the float->dt64 astype issue, i.e. unrelated

File: pandas/core/algorithms.py [hunk 2]
```diff
@@ -772,25 +752,13 @@ def factorize(
...
-            codes, uniques = values.factorize(  # type: ignore[call-arg]
-                use_na_sentinel=na_sentinel is not None
-            )
-        else:
-            na_sentinel_arg = -1 if na_sentinel is None else na_sentinel
-            with warnings.catch_warnings():
-                # We've already warned above
-                warnings.filterwarnings("ignore", ".*use_na_sentinel.*", FutureWarning)
-                codes, uniques = values.factorize(na_sentinel=na_sentinel_arg)
+        # GH#46910 - TimelikeOps has deprecated signature
```
Comment (jbrockmendel) [inline pandas/core/algorithms.py:755]: is it not deprecated there too?

File: pandas/core/arrays/base.py [hunk 3]
```diff
@@ -454,24 +450,6 @@ def __ne__(self, other: Any) -> ArrayLike:  # type: ignore[override]
         """
         return ~(self == other)
 
-    def __init_subclass__(cls, **kwargs) -> None:
```
Comment (jbrockmendel) [inline pandas/core/arrays/base.py:457]: wow didn't realize we had done this.  will be nice to see this go

File: pandas/core/algorithms.py (hunk 2 above)
Comment (rhshadrach) [reply pandas/core/algorithms.py]: Thanks, missed this one.

**Review by jorisvandenbossche — APPROVED** [review]
Looks good! This is indeed nice to see go ;)

**Comment by mroeschke** [comment]
@rhshadrach looks like there's a merge conflict; otherwise, looks good to merge

**Comment by mroeschke** [comment]
Thanks @rhshadrach

**Pull request merged**